### PR TITLE
osd: minimize the PG::get_osdmap()-driven ref-counting

### PR DESCRIPTION
**Pull request by rzarzynski**

### CODE REVIEW

File: src/osd/PrimaryLogPG.cc [hunk 1]
```diff
@@ -2020,7 +2020,8 @@ void PrimaryLogPG::do_op(OpRequestRef& op)
   }
 
   // blacklisted?
-  if (get_osdmap()->is_blacklisted(m->get_source_addr())) {
+  if (with_osdmap([&](const OSDMap& om) {
+        return om.is_blacklisted(m->get_source_addr());})) {
```
Comment (liewegas) [inline src/osd/PrimaryLogPG.cc]: can't we accomplish the same thing without a lambda if get_osdmap() returns OSDMapRef& ?

Comment (rzarzynski) [reply src/osd/PrimaryLogPG.cc]: I'm not really happy with the lambda design. It obfuscates `if`s while I haven't encoutered any complex appliance of `get_osdmap` yet which could benefit from it. This requires simplification.

On the other hand just changing the return type seems risky. What about replacing with `const OSDMap& get_osdmap_weakref() const`?

Comment (rzarzynski) [reply src/osd/PrimaryLogPG.cc]: The more I think, the less risk I see. Initially I was afraid about `auto`-based constructs, especially in the spot of merging old PRs (potential surprise after a rebase that doesn't fail compilation).

```cpp
{
  auto map_with_extended_life = pg->get_osdmap();
  do_later([=]() {
    // operate with assumption that the map won't disappear.
    // This should be safe as: 1) having a pg means also having
    // osdmap_ref and 2) https://stackoverflow.com/a/19676545
  });

  // ...
}
```

Maybe `decltype()`-based magic in struct definitions could be an issue but appears quite unlikely. Still, my imagination is pretty limited while the number of `get_osdmap()` occurrences is huge and the language is vast.

Comment (liewegas) [reply src/osd/PrimaryLogPG.cc]: I would just grep for get_osdmap() and auto and make sure we're explicitly assigning the result to OSDMapRef.. that's a quick check.

Comment (rzarzynski) [reply src/osd/PrimaryLogPG.cc]: `auto` and `get_osdmap` in same lines:
```
$ grep -n -r get_osdmap ../src/ --exclude=tags | grep auto
../src/tools/rebuild_mondb.cc:283:      const auto oid = OSD::get_osdmap_pobject_name(e);
../src/osd/OSD.cc:1665:  const auto epoch = pg->get_osdmap_epoch();
../src/osd/OSD.cc:9356:  auto curmap = sdata->get_osdmap();
../src/osd/PG.cc:1887:      auto& removed_snaps_queue = get_osdmap()->get_removed_snaps_queue();
../src/osd/PG.cc:2489:  auto state = get_osdmap()->get_state(osd->whoami);
```

Comment (liewegas) [reply src/osd/PrimaryLogPG.cc]: yep.. looks like ../src/osd/OSD.cc:9356 is the only one that really needs to be changed.

Comment (rzarzynski) [reply src/osd/PrimaryLogPG.cc]: It looks it's just a name collision:

```cpp
void OSD::dequeue_peering_evt(
  OSDShard *sdata,
  PG *pg,
  PGPeeringEventRef evt,
  ThreadPool::TPHandle& handle)
{
  PG::RecoveryCtx rctx = create_context();
  auto curmap = sdata->get_osdmap();
  // ...
}
```

We're calling `OSDShard::get_osdmap()` which isn't touched at all by this PR.
```cpp
struct OSDShard {
  // ...
  OSDMapRef shard_osdmap;

  OSDMapRef get_osdmap() {
    std::lock_guard l(osdmap_lock);
    return shard_osdmap;
  }
```

**Comment by yuriw** [comment]
[wip-yuri3-testing-2018-11-09-1730](https://trello.com/c/aXtu68eZ/470-wip-yuri3-testing-2018-11-09-1730)

**Pull request merged**